File: backend/src/tags/config_tag.py
import os
import re
from fastapi import HTTPException
from src.tags.base import BaseTag, TagParseResult
from src.config import CONFIG_DIR # Ensure CONFIG_DIR points to your backend config folder

class ConfigTag(BaseTag):
    """
    Handles [config "filename.py"/] or dynamic context configurations loader.
    """
    name = "config"
    PATTERN = re.compile(r'^\[config\s+(?:"(?P<path>[^"]+)"|(?P<var>\{[A-Za-z0-9_\.]+\}))/\]$')

    def parse(self, line: str, line_idx: int, ctx: dict) -> TagParseResult:
        m = self.PATTERN.match(line)
        if not m:
            return TagParseResult(consumed=False)
        
        path_token = m.group("path") if m.group("path") is not None else m.group("var")
        return TagParseResult(
            step={"type": "load_config", "file_path": path_token.strip()},
            consumed=True
        )

    def execute(self, step: dict, ctx: dict):
        if step.get("type") != "load_config":
            return None

        file_target = step["file_path"]
        # Resolve target module absolute path coordinates
        if not file_target.startswith("/"):
            absolute_path = os.path.join(CONFIG_DIR, file_target)
        else:
            absolute_path = file_target

        # Strict check loop before compiling modules bindings
        if not os.path.exists(absolute_path):
            raise HTTPException(
                status_code=422,
                detail={
                    "status": "CONFIG_FILE_MISSING_ERROR",
                    "message": f"Initialization config file script not found on server disk: {file_target}",
                    "details": f"Expected target location: {absolute_path}"
                }
            )

        # Import the dynamic config loader directly into active env
        from src.runtime import load_py_config
        load_py_config(absolute_path, ctx["env"])
        return None

File: src/lib/server/tags/config_tag.ts
import { BaseTag, TagParseResult, type ExecResult } from './base.js';
import type { Step, ExecContext, ParseContext } from '../types.js';
import { loadConfigFile } from '../config_loader.js';

/**
 * Handles [config "filename"/].
 *
 * The tag emits a `load_config` step. Actual module loading is async,
 * so the runtime awaits it separately (see runtime.ts).
 */
export class ConfigTag extends BaseTag {
    name = 'config';
    private readonly PATTERN = /^\[config\s+(?:"([^"]+)"|(\{[A-Za-z0-9_.]+\}))\s*\/?\]$/;

    parse(line: string, _lineIdx: number, _ctx: ParseContext): TagParseResult {
        const m = this.PATTERN.exec(line);
        if (!m) return new TagParseResult({ consumed: false });

        let filename = (m[1] ?? m[2]).trim();
        if (!filename.endsWith('.ts') && !filename.endsWith('.js')) {
            filename = `${filename}.ts`;
        }

        return new TagParseResult({
            step: { type: 'load_config', filename },
            consumed: true,
        });
    }

    execute(_step: Step, _ctx: ExecContext): ExecResult {
        // Handled asynchronously; see executeConfigStep below.
        return null;
    }
}

export async function executeConfigStep(step: Step, ctx: ExecContext): Promise<void> {
    if (step.type !== 'load_config') return;
    const filename = step.filename as string;
    await loadConfigFile(filename, ctx.env);
}